design.md

```diff
@@ -0,0 +1,138 @@
+  - ! RFC: Schema of a robot config file.
+    - Kinematics
+    - Device
+    - Device resources
+    - Scripts for body control & cad, limb control & cad, joint control
+    - Device resources in the robot config file need a flag that says whether they are allowed to fail discovery or not. By default, this is false, meaning that all device resources need to complete discovery without errors or else the entire discovery process fails. If it is true, and that device resources fails discovery, then the discovery process does not fail. This could be useful for competition robots where the user would rather the robot work with partial functionality rather than not work at all.
+- All hardware access must go through a robot (i.e., no direct access to hardware).
+  - Scripts would not work in a simulator if they could directly access the hardware.
+- We should use most of the hardware layer from the previous version of the kernel.
+
+#### RPC
+
+- We should use the Bowler RPC protocol from the previous version of the kernel.
+  - Need to add support for asking the device its capabilities so that the kernel can figure out if it needs to upload a new program.
+
+#### Dependency Management
+
+- The kernel must be able to download the toolchain for the device specified in the robot config file. This is because of the [usability requirement](#usability).
+  - The location the toolchain is downloaded from must be stored separately from the source code so that we can update it independently of the kernel. This must also include the version range of the kernel that the toolchain is compatible with. This should be implemented as a GitHub repo containing a registry of toolchains.
+
+#### uC Code
+
+- The Bowler RPC protocol can handle dynamic definition of device resources, but including support for every possible device resource is not possible because uC's have limited space. Therefore, the kernel must be able to generate, compile, and flash C++ programs to the device.
+- The kernel must be able to determine if it needs to upload a new program to the device. It will do this by asking the device for its capabilities, which will include the supported device resources (see [RPC](#rpc)). If a device resource in the robot config file being loaded is not in the supported capabilities, then the kernel must generate, compile, and flash a new program to the device.
+
+### CAD
+
+- Use the existing JCSG API for now.
+
+## Non-Functional Requirements
+
+### Open Source
+
+- All core components and their dependencies must be open-source under a compatible license. Some optional, non-core components may have closed-source dependencies.
+
+### Portability
+
+- x86_64 Windows, macOS, and Linux must be supported. Arm linux must be supported.
+
+### Responsiveness
+
+- The kernel's runtime must not pause for longer than 15 milliseconds when running under ZGC. Runtimes that must run with another garbage collector, like G1, do not have this requirement.
```

Comment:
Should we qualify this with certain minimum system requirements?

Reply:
I agree that there should be minimum system requirements somewhere, but I don't think here is the right place. I think that our website is a better place.